### ISSUE TITLE
Network traffic indicator not working 

### ISSUE THREAD

**Issue by Divyansh-2903** [mytja/treble_evo#79]
Is just shows 0 kb/s on the status bad and the data usage option is also broken , plus Im facing lags on this build while gaming , the UI is smooth but fps are quite unstable on this build , in my opinion A13 provide better performance than A14 , why is this so?
Also my processor is Mediatek dimensity 800U

**Comment by mytja** [comment]
> Is just shows 0 kb/s on the status bad and the data usage option is also broken

Is this on Android 15? Because it works for me on A14.

> in my opinion A13 provide better performance than A14 , why is this so?

You've answered yourself. Android is not becoming much more optimized over the years.

**Comment by Tanvir1337x** [comment]
Doesn't work for me either. Tried 9.6 and 9.5 Beta.

**Comment by Technical06** [comment]
> > Is just shows 0 kb/s on the status bad and the data usage option is also broken
> 
> Is this on Android 15? Because it works for me on A14.
> 
> > in my opinion A13 provide better performance than A14 , why is this so?
> 
> You've answered yourself. Android is not becoming much more optimized over the years.

i tried it on A14 still its broken , and about the performance when i compared the throttle test of superior os A13 and Evox 9.6 A14 
i found better results on A14 on paper but idk if A13 provides more performance compared to A14 ,i just cant say it now as i felt A13 was better but on scores A14 performs better and after watching the scores i think A14 is good too

**Comment by Technical06** [comment]
> > Is just shows 0 kb/s on the status bad and the data usage option is also broken
> 
> Is this on Android 15? Because it works for me on A14.
> 
> > in my opinion A13 provide better performance than A14 , why is this so?
> 
> You've answered yourself. Android is not becoming much more optimized over the years.

any solution for network traffic indicator not working?

**Comment by mytja** [comment]
Okay, thanks to everybody for feedback and sorry for the delay. It'd be of great help if anybody with a broken network traffic indicator would execute the following in the command prompt/appropriate OS shell and send me the output (make sure you have either mobile data or WiFi turned on & connected):

```
adb shell cat /sys/class/net/wlan0/statistics/rx_bytes
adb shell ifconfig
adb shell uname -a
```

Make sure you are on 2024-12-26 build or later, otherwise these commands won't work.

There should be nothing sensitive in those logs, so you can send them here on GitHub. If you still wish to keep things private, send them over to mytja@proton.me.

Thanks and have a nice day.

--------

NOTE FOR FUTURE DEVELOPERS

I've already tried switching to another API for obtaining the traffic data, however it only listed wrong traffic data ([also reported by others](https://forum.earlybird.club/threads/networkstatsmanager-querysummaryfordevice-strange-results.1305066/)). My implementation as a reference is available below in case anybody in the future would like to go ahead and improve on it (just rename it from .txt to .patch, as unfortunately GitHub does not allow for .patch files to be uploaded).

[0001-Alternative-broken-way-of-detecting-network-traffic.txt](https://github.com/user-attachments/files/18245802/0001-Alternative-broken-way-of-detecting-network-traffic.txt)

**Comment by Technical06** [comment]
> Okay, thanks to everybody for feedback and sorry for the delay. It'd be of great help if anybody with a broken network traffic indicator would execute the following in the command prompt/appropriate OS shell and send me the output (make sure you have either mobile data or WiFi turned on & connected):
> 
> ```
> adb root
> adb shell cat /sys/class/net/rmnet0/statistics/rx_bytes
> adb shell cat /sys/class/net/ppp0/statistics/rx_bytes
> adb shell cat /sys/class/net/rmnet0/statistics/tx_bytes
> adb shell cat /sys/class/net/ppp0/statistics/tx_bytes
> adb shell ls /sys/class/net/
> adb shell ifconfig
> adb shell uname -a
> ```
> 
> These commands can be executed on **any** GSI, so no need to flash Evolution X again just to test this out.
> 
> There should be nothing sensitive in those logs, so you can send them here on GitHub. If you still wish to keep things private, send them over to [mytja@proton.me](mailto:mytja@proton.me).
> 
> Thanks and have a nice day.
> 
> NOTE FOR FUTURE DEVELOPERS
> 
> I've already tried switching to another API for obtaining the traffic data, however it only listed wrong traffic data ([also reported by others](https://forum.earlybird.club/threads/networkstatsmanager-querysummaryfordevice-strange-results.1305066/)). My implementation as a reference is available below in case anybody in the future would like to go ahead and improve on it (just rename it from .txt to .patch, as unfortunately GitHub does not allow for .patch files to be uploaded).
> 
> [0001-Alternative-broken-way-of-detecting-network-traffic.txt](https://github.com/user-attachments/files/18245802/0001-Alternative-broken-way-of-detecting-network-traffic.txt)

Im uploading the logs here , idk if they will help u or not
[NTIlogs.txt](https://github.com/user-attachments/files/18246961/NTIlogs.txt)

**Comment by mytja** [comment]
Thanks, that helps me greatly. Do you need a slim or a full build? I'll try building with the fixes now.

**Comment by mytja** [comment]
The 2024-12-25 release has been updated with new builds. Please check them out & report back if it works or not.

**Comment by SMOLKEYS** [comment]
Got this issue myself on Andy's LOS 21 with a Helio G95 chipset. Gonna test this.

**Comment by mdsyfulislam3249** [comment]
Now I'm facing only the Network Traffic Indicator issue. Because of that Data Usage is also not working. Here is the SS. MediaTek G85 | Kernel: 4.14. I flashed the latest build of Evo v9.7 - 25 Dec on a MediaTek G88 device. Everything works fine.

![image](https://github.com/user-attachments/assets/c1fbd3ee-645f-4b98-8875-7b98247be925)

**Comment by mytja** [comment]
Thanks, but you've cut off the last command. Any chance of sending that one, as it's the most important one?

**Comment by mdsyfulislam3249** [comment]
![image](https://github.com/user-attachments/assets/073c4fe6-6a19-4b7c-8d3d-176dce9ef328)
![image](https://github.com/user-attachments/assets/179128d6-03ec-4ad5-a055-570315da9b7a)

**Comment by mytja** [comment]
Uhh, you have the latest build and it doesn't work?

Would you mind executing the following command and sending the output?

```
adb logcat | findstr -i NetworkTraffic
```

**Comment by mdsyfulislam3249** [comment]
![image](https://github.com/user-attachments/assets/2e8ff233-ff61-466c-ae11-d796e109c2a2)

Just Looping

**Comment by mytja** [comment]
Thank you so much. I'll send a test build once I finish debugging it. Would you like a slim build or a normal build?

**Comment by mdsyfulislam3249** [comment]
I love slim build. Build this version then.

**Comment by mytja** [comment]
I've decided to build both versions, as I've included additional fixes for some other issues. I've tested the fix for this issue myself, and it seems to be working properly (it should work for WiFi, I'm unsure for mobile data). Let me know if it works for you.

I'll take a break from building now, so you can't expect fixes anytime soon. As said, it should work properly. If it doesn't, please comment and I'll reopen the issue.

https://github.com/mytja/treble_evo/releases/tag/20241226

**Comment by mdsyfulislam3249** [comment]
Thanks a lot dude. I will try it ASAP and let you know the feedback.

**Comment by SMOLKEYS** [comment]
Currently having the same issue here. I went and tested the latest Dec 26 build and it didn't fix the issue the original reporter had.

**Comment by mytja** [comment]
> Currently having the same issue here. I went and tested the latest Dec 26 build and it didn't fix the issue the original reporter had.

Could you send the output of these commands?

```
adb shell cat /sys/class/net/wlan0/statistics/rx_bytes
adb shell ifconfig
adb shell uname -a
```

-----

I've hidden some comments in order to increase the visibility while scrolling this comment thread. Don't mind the "outdated" tag, I couldn't find a better option.

**Comment by SMOLKEYS** [comment]
RX output: `843894`
TX output: `309484`
uname output: `Linux localhost 4.14.186-gfc8c20b58a12-dirty #1 SMP PREEMPT Tue Oct 19 13:35:50 CST 2021 aarch64 Android`

ifconfig output:
```
ccmni1: flags=193<UP,RUNNING,NOARP>  mtu 1500
        inet 10.124.0.82  netmask 255.0.0.0
        unspec 12-27-7C-4A-8E-E3-00-00-00-00-00-00-00-00-00-00  txqueuelen 1000  (UNSPEC)
        RX packets 68611  bytes 74968590 (71.4 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 45392  bytes 6621057 (6.3 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

dummy0: flags=195<UP,BROADCAST,RUNNING,NOARP>  mtu 1500
        inet6 fe80::38f5:2fff:feaf:893a  prefixlen 64  scopeid 0x20<link>
        ether 3a:f5:2f:af:89:3a  txqueuelen 1000  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 4648  bytes 775618 (757.4 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 4094  bytes 774070 (755.9 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 4094  bytes 774070 (755.9 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

wlan0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        ether be:d7:ad:53:fe:55  txqueuelen 3000  (Ethernet)
        RX packets 1333  bytes 843894 (824.1 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 1597  bytes 309484 (302.2 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

**Comment by Technical06** [comment]
> adb shell uname -a

![image](https://github.com/user-attachments/assets/d02e3028-dd8d-4bdd-a7f0-d4ef8e12179c)
![image](https://github.com/user-attachments/assets/62f016d8-5772-4d29-8e85-1e19aecda78f)

the bluetooth issue and network traffic issue is still not fixed for me

**Comment by Technical06** [comment]
> > Currently having the same issue here. I went and tested the latest Dec 26 build and it didn't fix the issue the original reporter had.
> 
> Could you send the output of these commands?
> 
> ```
> adb shell cat /sys/class/net/wlan0/statistics/rx_bytes
> adb shell ifconfig
> adb shell uname -a
> ```
> 
> I've hidden some comments in order to increase the visibility while scrolling this comment thread. Don't mind the "outdated" tag, I couldn't find a better option.


![image](https://github.com/user-attachments/assets/d02e3028-dd8d-4bdd-a7f0-d4ef8e12179c)
![image](https://github.com/user-attachments/assets/62f016d8-5772-4d29-8e85-1e19aecda78f)

the bluetooth issue and network traffic issue is still not fixed for me

**Comment by mytja** [comment]
Ughh. Can you please send the output of `adb logcat | findstr -i NetworkTraffic`? I'm completely lost, as I've tested this thing myself.

**Comment by SMOLKEYS** [comment]
> Ughh. Can you please send the output of `adb logcat | findstr -i NetworkTraffic`? I'm completely lost, as I've tested this thing myself.

[10s.txt](https://github.com/user-attachments/files/18259624/10s.txt)

~~Also, probably irrelevant to here, but I also have an issue related with Bluetooth. [You can go here for the issue report.](https://github.com/TrebleDroid/treble_experimentations/issues/127) It's actually fixed, but the patch wasn't backported to A14.~~

**Comment by mytja** [comment]
Thanks. What about `adb logcat | findstr -i denied`?

Will do the Bluetooth backporting in the January build, thanks.

**Comment by Technical06** [comment]
> Thanks. What about `adb logcat | findstr -i denied`?
> 
> Will do the Bluetooth backporting in the January build, thanks.

ill send it , just wait 5 minutes
ill attach it here
[networktraffic.txt](https://github.com/user-attachments/files/18259816/networktraffic.txt)
[adb-logcat-findstr-i-denied.txt](https://github.com/user-attachments/files/18259851/adb-logcat-findstr-i-denied.txt)

**Comment by Technical06** [comment]
> > Ughh. Can you please send the output of `adb logcat | findstr -i NetworkTraffic`? I'm completely lost, as I've tested this thing myself.
> 
> [10s.txt](https://github.com/user-attachments/files/18259624/10s.txt)
> 
> ~Also, probably irrelevant to here, but I also have an issue related with Bluetooth. [You can go here for the issue report.](https://github.com/TrebleDroid/treble_experimentations/issues/127) It's actually fixed, but the patch wasn't backported to A14.~

Thats the same issue with me , i also have mediatek Chipset and this is the same issue ive been facing

**Comment by mytja** [comment]
@Technical06, thank you so much for your logs. This should give me enough information to fix it. SELinux is hell.

Despite me saying that I'll take break, I'm so invested into this issue that you'll receive a test image in a couple of hours.

**Comment by mytja** [comment]
Please test this new image (it should also contain the backported Bluetooth fix):

https://git.severkar.eu/mytja/treble_evo_test/releases/tag/20241227-test1-traffic-bt

**Comment by Technical06** [comment]
> Please test this new image (it should also contain the backported Bluetooth fix):
> 
> https://git.severkar.eu/mytja/treble_evo_test/releases/tag/20241227-test1-traffic-bt

okay sure

**Comment by SMOLKEYS** [comment]
> Please test this new image (it should also contain the backported Bluetooth fix):
> 
> https://git.severkar.eu/mytja/treble_evo_test/releases/tag/20241227-test1-traffic-bt

Tested the build. The network traffic indicator now works (except for mobile data, for some reason), but the [aforementioned Bluetooth issue](https://github.com/TrebleDroid/treble_experimentations/issues/127) still persists.

**Comment by Technical06** [comment]
> > Please test this new image (it should also contain the backported Bluetooth fix):
> > https://git.severkar.eu/mytja/treble_evo_test/releases/tag/20241227-test1-traffic-bt
> 
> Tested the build. The network traffic indicator now works (except for mobile data, for some reason), but the [aforementioned Bluetooth issue](https://github.com/TrebleDroid/treble_experimentations/issues/127) still persists.

Nice , the download speed was exceptionally Slow for me though

**Comment by mytja** [comment]
That's great hearing that the network indicator now works. As mentioned somewhere in the thread before, only WiFi should be monitored. I'll provide another test build with fixed mobile data monitoring, should be easy now. Apart from that, thanks for your responsiveness & quick testing.

Regarding the Bluetooth patches, I'll do a deeper dive in the January build.

**Comment by SMOLKEYS** [comment]
Nice. One of the primary reasons for my responsiveness was because of the _egregious_ amount of suffering I had to go through because of Bluetooth.

In my year of using GSIs on three devices, Bluetooth was _**always**_ the one part of them that just had one issue or another that would make it tedious to use or outright impossible.

**Comment by Technical06** [comment]
> That's great hearing that the network indicator now works. As mentioned somewhere in the thread before, only WiFi should be monitored. I'll provide another test build with fixed mobile data monitoring, should be easy now. Apart from that, thanks for your responsiveness & quick testing.
> 
> Regarding the Bluetooth patches, I'll do a deeper dive in the January build.

actually just for your knowledge , bluetooth works on all A13 builds as i told you , and old A14 builds which are now legacy 
maybe it may not be possible for you to use those patches because they're old , but you can try

**Comment by SMOLKEYS** [comment]
Oh, and might just be me, but data/wifi usage tracking doesn't seem to be working too.

**Comment by SMOLKEYS** [comment]
![Screenshot_20241228-020021_Settings](https://github.com/user-attachments/assets/a524d460-2670-4cb6-8224-5c90ff795158)

(this lil fella in the quick settings)

addr:

Oh, and another more minor but inconvenient issue: DSU doesn't work with an unencrypted data partition. PHH provided me a test build which fixes that, but I couldn't test it and I'm also not sure what patches he applied to get that.

**Comment by Technical06** [comment]
> ![Screenshot_20241228-020021_Settings](https://private-user-images.githubusercontent.com/75618732/398987752-a524d460-2670-4cb6-8224-5c90ff795158.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUzMjM4MzUsIm5iZiI6MTczNTMyMzUzNSwicGF0aCI6Ii83NTYxODczMi8zOTg5ODc3NTItYTUyNGQ0NjAtMjY3MC00Y2I2LTgyMjQtNWM5MGZmNzk1MTU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI3VDE4MTg1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc5YjU2OWE1YzZmMWRmMmUyMzZiNjVhOWJkODM2NWI1MmZjNmQ4OTJhYTQzOWFkODcwNjZjZDZmOTFlMzA2YjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.LSrmkOeLvFSM29qWKRaX1UBHAo3XyPrs9p7uQn9k-kM)
> 
> (this lil fella in the quick settings)
> 
> addr:
> 
> Oh, and another more minor but inconvenient issue: DSU doesn't work with an unencrypted data partition. PHH provided me a test build which fixes that, but I couldn't test it and I'm also not sure what patches he applied to get that.

Same goes for me , even i cant use the Usage settings and also DSU doesnt work for me , though my device is encrypted

**Comment by mytja** [comment]
> @SMOLKEYS: Nice. One of the primary reasons for my responsiveness was because of the egregious amount of suffering I had to go through because of Bluetooth.
>
> In my year of using GSIs on three devices, Bluetooth was always the one part of them that just had one issue or another that would make it tedious to use or outright impossible.

Yeah, I get you. Bluetooth on Android is a mess... well, everything is a mess. I & most other builders most frequently get reports of FOD & Bluetooth not working. It's especially annoying, as most experience their issue on ALL GSIs, but don't report it on TrebleDroid. I can't offer knowledgeable support on Bluetooth and FOD. I tried fixing FOD on A15 for my device, but ultimately failed.

> @mytja: Regarding the Bluetooth patches, I'll do a deeper dive in the January build.

I've decided to redo the patches right now. I don't know if it'll do much, as TrebleDroid Bluetooth patches apply more or less cleanly, meaning that Lineage hasn't changed much.

> @SMOLKEYS: Oh, and might just be me, but data/wifi usage tracking doesn't seem to be working too.

Yeah, under the hood Android is using a very different API. Your old kernel doesn't offer support for these functionalities, therefore nothing is reported. Unfortunately, my fixes are great for the traffic indicator, but not measuring traffic over time.

I've disabled this as a temporary measure, so you won't see it in the next build. I have an idea on how to fix it however.

Note to self: save traffic indicator data to SystemProperties in `frameworks/base/packages/SystemUI/src/com/android/systemui/statusbar/NetworkTraffic.java` & read it from `frameworks/base/packages/SystemUI/src/com/android/systemui/qs/QSFooterView.java`. Figure out how to save this even when traffic indicator is disabled.

> @SMOLKEYS: Oh, and another more minor but inconvenient issue: DSU doesn't work with an unencrypted data partition. PHH provided me a test build which fixes that, but I couldn't test it and I'm also not sure what patches he applied to get that.

I'll sure implement it if phh decides to include it in TrebleDroid's A15 patchset. For now I can't do much. I have encrypted data and wouldn't like to decrypt it just to test this.

> @Technical06: DSU doesnt work for me , though my device is encrypted

This likely means that your vendor doesn't properly support DSU. That's unfortunate & I can't do much about it even if you open an issue.

-------

Anyhow, this issue has turned into such a mess with (now) 39 comments. I'll post a new test build here (I'll leave it building overnight, so expect it tomorrow), but for other issues I moving to #104 for the Bluetooth issue and creating a new issue for that "data usage over the entire day" issue.

Thanks to everyone in this thread for testing & reporting. Have nice holidays.

**Comment by Technical06** [comment]
Sorry i think my explanation wasn't enough , DSU works for me in A13 builds but it shows some error failed to create partition on A14 builds . this only happens now on the Recent builds .

**Comment by mytja** [comment]
> Sorry i think my explanation wasn't enough , DSU works for me in A13 builds but it shows some error failed to create partition on A14 builds . this only happens now on the Recent builds .

That's an interesting bug, and I don't know why it would fail...

---------

New build, which should fix traffic indicator on mobile networks & maybe Bluetooth:

https://git.severkar.eu/mytja/treble_evo_test/releases/tag/20241228-test1-traffic-bt OR
https://github.com/mytja/treble_evo_test/releases/tag/20241228-test1-traffic-bt

**Comment by SMOLKEYS** [comment]
Noted, testing will now begin.

**Comment by mdsyfulislam3249** [comment]
> > Sorry i think my explanation wasn't enough , DSU works for me in A13 builds but it shows some error failed to create partition on A14 builds . this only happens now on the Recent builds .
> 
> That's an interesting bug, and I don't know why it would fail...
> 
> New build, which should fix traffic indicator on mobile networks & maybe Bluetooth:
> 
> https://git.severkar.eu/mytja/treble_evo_test/releases/tag/20241228-test1-traffic-bt OR https://github.com/mytja/treble_evo_test/releases/tag/20241228-test1-traffic-bt

![image](https://github.com/user-attachments/assets/95634438-7ef0-4675-998b-9c845e2b9f0e)
Traffic Indicator on Mobile Network is still not working. Wifi is working fine. Everything else so smooth. Thanks for your effort dear.

**Comment by mytja** [comment]
Download this image:

https://github.com/mytja/treble_evo_test/releases/tag/20241230-test3-bt-charger

It fixes the issue you're having.

**Comment by mdsyfulislam3249** [comment]
> Download this image:
> 
> https://github.com/mytja/treble_evo_test/releases/tag/20241230-test3-bt-charger
> 
> It fixes the issue you're having.

Thanks a lot, dear, it's working fine.

**Comment by mdsyfulislam3249** [comment]
Hi, Mytja. Hope you are doing well. I flashed your Evo X 10.2 in my phone's kernel: 4.14.186, G85 chipset. the issues I have found are no internet (wifi connected but the internet is not working for unable IP address). The fingerprint icon goes outside of the screen and looks like a big icon. Other than everything is so smooth and all are ok.

But I flash this from in my other phone kernel: 4.19.191.  G88 chipset. I haven't found a single bug. Everything works fine and smooth UI experience.

If you can fix it for kernel 4.14.186, please help me to fix these issues. Please make a test build for me.

Love you for your kind services.
